File: internal/cli/testdata/golden/kitchen-sink/api/_cloudcc_runtime/kv.py
"""Key/value store backed by DynamoDB.

Values are stored JSON-encoded under a single ``value`` attribute so that any
JSON-shaped dict round-trips unchanged. Storing the dict as native DynamoDB
attributes instead would quietly turn numbers into Decimals and reject empty
values, which is not what the local emulation does.
"""

import json

from . import _client


def connect(id):
    """Return a client for the KV store declared as ``persist(KVStore(), id=...)``."""
    name = _client.env("CLOUDCC_KV_%s_TABLE" % _client.slug(id), "persist", id)
    return KVStore(id, _client.resource("dynamodb").Table(name))


class KVStore:
    def __init__(self, id, table):
        self.id = id
        self._table = table

    def get(self, key):
        """Return the item at ``key``, or None."""
        item = self._table.get_item(Key={"id": str(key)}).get("Item")
        if item is None:
            return None
        return json.loads(item["value"])

    def put(self, key, value):
        """Store ``value`` at ``key``."""
        self._table.put_item(Item={"id": str(key), "value": json.dumps(value)})

    def delete(self, key):
        """Remove ``key`` if present."""
        self._table.delete_item(Key={"id": str(key)})

    def keys(self):
        """Every key currently stored, sorted."""
        out = []
        kwargs = {"ProjectionExpression": "id"}
        while True:
            page = self._table.scan(**kwargs)
            out.extend(item["id"] for item in page.get("Items", []))
            token = page.get("LastEvaluatedKey")
            if not token:
                return sorted(out)
            kwargs["ExclusiveStartKey"] = token

    def __repr__(self):
        return "<KVStore %r (dynamodb)>" % self.id

File: internal/lang/node/templates/_cloudcc_runtime/kv.js
// Key/value store backed by DynamoDB.
//
// Values are stored JSON-encoded under a single `value` attribute so that any
// JSON-shaped object round-trips unchanged. Storing native attributes instead
// would quietly reject empty strings and reorder nothing usefully.

import { DynamoDBClient, DeleteItemCommand, GetItemCommand, PutItemCommand, ScanCommand } from "@aws-sdk/client-dynamodb";

import { common, env, slug } from "./client.js";

export function connect(id) {
  const table = env(`CLOUDCC_KV_${slug(id)}_TABLE`, "persist", id);
  return new KVStore(id, table, new DynamoDBClient(common()));
}

export class KVStore {
  constructor(id, table, client) {
    this.id = id;
    this._table = table;
    this._client = client;
  }

  async get(key) {
    const out = await this._client.send(
      new GetItemCommand({ TableName: this._table, Key: { id: { S: String(key) } } }),
    );
    if (!out.Item) {
      return null;
    }
    return JSON.parse(out.Item.value.S);
  }

  async put(key, value) {
    await this._client.send(
      new PutItemCommand({
        TableName: this._table,
        Item: { id: { S: String(key) }, value: { S: JSON.stringify(value) } },
      }),
    );
  }

  async delete(key) {
    await this._client.send(
      new DeleteItemCommand({ TableName: this._table, Key: { id: { S: String(key) } } }),
    );
  }

  async keys() {
    const out = [];
    let start;
    do {
      const page = await this._client.send(
        new ScanCommand({
          TableName: this._table,
          ProjectionExpression: "id",
          ExclusiveStartKey: start,
        }),
      );
      for (const item of page.Items ?? []) {
        out.push(item.id.S);
      }
      start = page.LastEvaluatedKey;
    } while (start);
    return out.sort();
  }
}
